packages/next/client/page-loader.js

```diff
@@ -70,14 +69,54 @@ export default class PageLoader {
         return
       }
 
-      // Load the script if not asked to load yet.
       if (!this.loadingRoutes[route]) {
-        this.loadScript(route)
+        // Make sure we don't load a dependency that's already loaded
+        const loadedModules = new Set(
+          Array.from(document.getElementsByTagName('script'), el => {
+            const results = /\/_next\/.*/g.exec(el.src)
+            return results ? results[0] : el.src
+          })
+        )
+        const deps = this.getDependencies(route)
+        const urlsToLoad = deps
+          .map(url => '/_next/' + url)
+          .filter(url => !loadedModules.has(url))
+
+        // Wait for all of the page's dependencies to load before loading
```

Comment:
From side discussion with @spanicker - maybe consider using preload to start fetching the route chunk in parallel with its dependencies here?

```js
let dependenciesReady = Promise.resolve();
if (urlsToLoad.length > 0) {
  this.preload(route);
  dependenciesReady = Promise.all(urlsToLoad.map(this.loadDependency))
}
this.loadingRoutes[route] = true
dependenciesReady.then(() => {
  this.loadScript(route)
})
```

Reply:
I'd like to save this for an optional optimization we can implement after seeing how this performs on a real site. Currently Next prefetches all Links except those which are explicitly excluded, which I have to imagine is a fairly small number of them. Additionally, we'd have to write a separate prefetch routine just for this case, as the current prefetch method falls back to just straight up downloading and executing the script, which obviously doesn't make sense to do in this situation

Reply:
Agreed. One thing we never touched on yesterday though: does it actually matter if the route is loaded before its dependencies?  For standard webpack chunk loading, the bundled code just registers itself with webpack, and execution of the chunk's modules waits for dependencies to be loaded (done in webpack's internal chunk loading). 
Next circumvents the chunk loading, but it is still using the module to chunk resolution? If so then it might be okay to load the route first and let it sit pending while its deps finish.

Reply:
Updated this, thanks for the tip @developit!